### ISSUE TITLE
False alarm TRIO107 with try/finally block

### ISSUE THREAD

**Issue by Zac-HD** [Zac-HD/flake8-trio#45]
```python
async def to_queue(iter_func, queue):
    async with iter_func() as it:
        async for x in it:
            queue.put(x)


async def to_queue2(iter_func, queue):
    try:
        async with iter_func() as it:
            async for x in it:
                queue.put(x)
    finally:
        queue.put(None)
```
```
$ flake8 example.py
example.py:7:1: TRIO107: exit from async function with no guaranteed checkpoint
                         or exception since function definition on line 7
```

but if the first case is fine, the second should be fine too.

**Comment by Zac-HD** [comment]
```python
async def f():
    with ...:
        try:
            await trio.sleep(0)
        finally:
            ...
```

Seems like a simpler repro, probably the same underlying issue?

**Comment by jakkdl** [comment]
(the `with` is also irrelevant)

Took me a while to figure out wth is going on here, and ultimately the error was due to me wanting to catch
```python
try:
  [code that might error]
  await trio.sleep(0)
finally:
  return # or yield
```
So I always entered `finally` with a worst-case jump from `try` - which is correct for checking the content of the finalbody, but in the case of no errors it's incorrect to exit finally with that worst-case.

I also realized #48 while doing this.

**Comment by jakkdl** [comment]
looking more at program flow there's some silly dumb stuff if you do
```python
try:
  ...
except:
  yield
  raise
finally:
  yield # won't checkpoint after the yield in except when jumped to upon except re-raising
```
But not gonna cover that.

**Comment by jakkdl** [comment]
Fixed with #49
